tuf/ngclient/updater.py

```diff
@@ -0,0 +1,579 @@
+        updated_targets = []
+        updated_targetpaths = []
+
+        for target in targets:
+            # Prepend 'destination_directory' to the target's relative filepath
+            # (as stored in metadata.)  Verify the hash of 'target_filepath'
+            # against each hash listed for its fileinfo.  Note: join() discards
+            # 'destination_directory' if 'filepath' contains a leading path
+            # separator (i.e., is treated as an absolute path).
+            filepath = target["filepath"]
+            target_filepath = os.path.join(destination_directory, filepath)
+
+            if target_filepath in updated_targetpaths:
+                continue
+
+            # Try one of the algorithm/digest combos for a mismatch.  We break
+            # as soon as we find a mismatch.
+            for algorithm, digest in target["fileinfo"].hashes.items():
+                digest_object = None
+                try:
+                    digest_object = sslib_hash.digest_filename(
+                        target_filepath, algorithm=algorithm
+                    )
+
+                # This exception will occur if the target does not exist
+                # locally.
+                except sslib_exceptions.StorageError:
+                    updated_targets.append(target)
+                    updated_targetpaths.append(target_filepath)
+                    break
+
+                # The file does exist locally, check if its hash differs.
+                if digest_object.hexdigest() != digest:
+                    updated_targets.append(target)
+                    updated_targetpaths.append(target_filepath)
+                    break
+
+        return updated_targets
+
+    def download_target(
+        self,
+        targetinfo: Dict,
+        destination_directory: str,
+        target_base_url: Optional[str] = None,
+    ):
+        """
+        Download target specified by 'targetinfo' into 'destination_directory'.
+
+        Args:
+            targetinfo: data received from get_one_valid_targetinfo()
+            destination_directory: existing local directory to download into.
+                Note that new directories may be created inside
+                destination_directory as required.
+            target_base_url: Optional; Base URL used to form the final target
+                download URL. Default is the value provided in Updater()
+
+        Raises:
+            TODO: download-related errors
+            TODO: file write errors
+        """
+        if target_base_url is None and self._target_base_url is None:
+            raise ValueError(
+                "target_base_url must be set in either download_target() or "
+                "constructor"
+            )
+        if target_base_url is None:
+            target_base_url = self._target_base_url
+        else:
+            target_base_url = _ensure_trailing_slash(target_base_url)
+
+        full_url = parse.urljoin(target_base_url, targetinfo["filepath"])
```

Comment:
Per [5.7.3](https://theupdateframework.github.io/specification/v1.0.19/#fetch-target) of the detailed client workflow, if consistent snapshots are enabled for the repository then we should prefix the target's filename with the hash(es) when downloading.

Reply:
oops! Filed #1479